[mono] Add mono_class_get_special_jit_flags and MONO_SPECIAL_JIT_DEFER_CLASS_INIT_TO_CTOR

lambdageek opened the pull request:
This is a hack to address https://github.com/dotnet/runtime/issues/60550 and #60976 

Add a flag that the loader can set on a type to tell the JIT to delay initialization of a class mentioned in a `newobj` instruction until runtime.

Set the flag when we load `Xunit.ConsoleClient.ConsoleRunner` from xunit.console.dll
becuase it essentially has this in its Main:
```csharp
  AssemblyLoadContext.Default.Resolving += SomeResolvingHelper();
  new ConsoleRunner().EntryPoint (args);
```

So the call to `new ConsoleRunner` triggers class loading at JIT time to initialize the fields of `ConsoleRunner` before the resolving helper has been installed.

---

Also update the interpreter to respect the defer flag.

---

Also stop looking in `ApplicationBase` for assemblies.  (That's the #60976 part)

lambdageek committed:
[metadata] Add "special jit flags" to MonoClass

Add a rarely used MonoClass property for "special jit
flags" (`m_class_has_special_jit_flags` and `mono_class_get_special_jit_flags`)
that is set at class loading time on some classes known to the runtime.

Add an `enum MonoSpecialJitClassFlags` with the various values.  Currently
there's a single value `MONO_SPECIAL_JIT_USE_ICALL_NEWOBJ`.

When loading `Xunit.ConsoleClient.ConsoleRunner` from `xunit.console.dll`, set
the `USE_ICALL_NEWOBJ` flag.

The JIT handling of this flag is not yet implemented.

The `USE_ICALL_NEWOBJ` should tell the JIT to avoid initializing the class at
JIT-time, and to delegte new instance creation to an icall that will initialize
the class, allocate the memory and invoke the constructor.

This is part of a workaround for https://github.com/dotnet/runtime/issues/60550

lambdageek committed:
[mini] Add a jit icall for newobj construction

Avoid calling `mono_class_init_internal` for classes marked with
`MONO_SPECIAL_JIT_USE_ICALL_NEWOBJ` until runtime, rather than JIT-time.

Related to https://github.com/dotnet/runtime/issues/60550

lambdageek committed:
[loader] Don't proble ApplicationBase when resolving assemblies

Related to https://github.com/dotnet/runtime/issues/60976

lambdageek committed:
[mini] Skip class init before calls for special classes

lambdageek committed:
rename flag to MONO_SPECIAL_JIT_DEFER_CLASS_INIT_TO_CTOR

Make the name match the intent: deferring class initialization until an
instance must be created.

lambdageek committed:
[interp] defer class initialization for special classes

jkotas commented:
We consider the exact timing of the static assembly loads to be runtime implementation detail that is subject to change. If we happened to change the CoreCLR behavior to be more like Mono for some reason, we would consider such change to be "by design".

From https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/loading-managed#when-are-managed-assemblies-loaded : "The exact timing of when the static assembly references are loaded is unspecified. It can vary between runtime versions and it is influenced by optimizations like inlining."

lambdageek commented:
hmm... not probing `ApplicationBase` in Mono breaks tests.  But I can't find anything in CoreCLR probing logic that uses it.  Strange

lambdageek commented:
> We consider the exact timing of the static assembly loads to be runtime implementation detail that is subject to change. If we happened to change the CoreCLR behavior to be more like Mono for some reason, we would consider such change to be "by design".

Aside from hardcoding xunit.console (which should probably just be fixed upstream), I'm now wondering whether doing less class initialization in Mono during JITing might improve startup on Android instead of initializing all classes on all branches in some method, we would only initialize the ones actually instantiated on some execution path.

uweigand commented on src/mono/mono/metadata/appdomain.c:
```diff
@@ -725,7 +725,8 @@ mono_domain_assembly_preload (MonoAssemblyLoadContext *alc,
 
-	if (!mono_runtime_get_no_exec ()) {
+        // only look in ApplicationBase for satellite assemblies
+	if (!mono_runtime_get_no_exec () && !mono_assembly_name_culture_is_neutral (aname)) {
```
Why the satellite assembly special case?   As far as I can tell, CoreCLR doesn't use ApplicationBase for those either ...

lambdageek replied:
Yea - that was wrong. I was guessing based on the test failures.

lambdageek committed:
Update the TPA list preload hook to check for .exe files, too

If the TPA list includes any .exe files, also try .exe extension on the
assembly.

Refactoring: combine all the loader properties into a single static struct.
Some of it comes from host properties (or pre-parsed properties), and some of
it (`tpa_has_exe`) is computed cached information.

lambdageek committed:
don't use non-standard strcasestr

lambdageek committed:
[mini][jit] Defer class initialization for nearly all constructors

In cases when we're not using generic sharing, and the type is a reference
type, just JIT as a call to a runtime helper and pass it the ctor token.  If
the constructor is on a branch that isn't taken, we avoid initializing a class
that may not be needed yet.

lambdageek commented:
/azp run runtime-manual

lambdageek commented:
/azp run runtime-staging-manual
